### DISCUSSION TITLE
Get return of a query from backend

### DISCUSSION THREAD

**Discussion by domjesus** [vercel/next.js#15066] [Help] [answered]
I'm trying to recover the result of sql query of backend, to server through getServerSideProps, but the result is just undefined. The object returns mounted like this {name:'name',id:'id'} Here's my code. I think is scope problem, but i don't find solution.

export async function getServerSideProps(context) {    

    const con = await connection();

   const sql = "select id, username from users";

    let employes = []

await query(con,sql, function(result){

      console.log(result) //works

      employes = result

     console.log(employes) // works fine

    })      
    
    console.log(employes); //NOT WORKS, IS UNDEFINED   

return {

      props: {        

        employes:employes

      }
    }

Full code: https://github.com/domjesus/learning-next-js/

Thanks for help.

**Comment by jamesmosier** [comment, marked as answer]
You are running into how JavaScript's closures work. When `function(result){}` is called, it is in a different scope than your `employees` variable. 

What does the code for `query()` look like?

**Reply by domjesus** [reply]
Hi, the full code is here: https://github.com/domjesus/learning-next-js/blob/master/utils/query.js, the project is https://github.com/domjesus/learning-next-js/.
Thanks

**Reply by jamesmosier** [reply]
Give this a try:

```js
// query.js
export async function query(connection,queryString) {
  return new Promise(function(resolve, reject) {
      // The Promise constructor should catch any errors thrown on
      // this tick. Alternately, try/catch and reject(err) on catch.

      connection.query(queryString, function (err, rows, fields) {
          if (err) {
            return reject(err);
          }
          resolve(rows);
      });
  });
}


// pages/servidores.js
export async function getServerSideProps(context) {
  try {
    const con = await connection();
    
    let servidores = [ 
        {nome:"jose",matricula:"123"},
        {nome:"maria",matricula:"12321"},
        {nome:"tonho",matricula:"994561"}
    ];

    const sql = "select MATRICULA as matricula,USER_NOME as user_nome from users order by user_nome DESC limit 33";

    const result = await query(con, sql);
    const serverTo = result;

    return {
      props: {
        serverTo
      }
    }
  } catch (e) {
    console.error(e)
  }
}
```

Now `query.js` returns a promise that you can `await` in `getServerSideProps`. `query()` will return the result of the sql query or throw an error if it fails.

**Reply by domjesus** [reply]
Hi, works fine. Thank you.
